enhancements/windows-containers/windows-machine-config-operator-upgrades.md

```diff
@@ -0,0 +1,206 @@
+
+The reason we need to cause the VMs to be fully terminated, instead of updating
+them, is that changes made to the VM done in previous versions of the WMCO may
+cause issues with the current version. If we chose to do an in place update,
+we would need to actively undo changes made to the node that may cause issues
+with the current WMCO version. The logic to keep track and undo these changes
+adds an additional layer of complexity that also adds the benefit of a quicker
+upgrade time. The risk of introducing additional issues is not worth it at this
+time, however it should be explored as the product matures.
+
+The reason we will not handle Windows operating system updates is that the
+cluster administrator provides the Windows image to create the VMs with, and
+it stays their responsiblity to provide an updated image.
+
+It is left to the user to create the Windows MachineHealthChecks, as
+WMCO's job is solely to configure Windows VMs into nodes. It is the user's
+responsibility to create the MachineSet which in turn creates the Windows
+Machines (VMs). Because of that, the responsibility falls onto the user to set
+up MachineHealthChecks, as WMCO does not create the Machines, and should not
+decide when they should be destroyed.
+
+### Design Details
+
+WMCO is published to OperatorHub as a Red Hat operator. When a new WMCO version
+is released, and the new version is compatible with the current cluster
+version, an update can occur. While we cannot explicity specify a required OCP
+version for the WMCO upgrade, each new OCP version upgrade should include a
+Kubernetes version upgrade. This allows us to use the [minKubeVersion](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/building-your-csv.md#operator-metadata)
+field in WMCO's CSV, which stops the operator from upgrading if the next
+WMCO version does not have at least the specified Kubernetes version.
+
+This update will either happen automatically, or require approval, based on
+the settings given through OLM when initially installing the operator.
+
+In order to facilitate upgrades, WMCO will add an annotation to the node
+indicating the version of the operator which configured the node.
+Using a node annotation is the easiest way to track this information in a way
+that lasts across WMCO restarts. WMCO is directly responsible for the creation
+of Windows nodes, and we should expect that other entities will not clear or
+alter the annotations WMCO gives to such nodes.
+
+On reconcile, if the annotated version does not match the current version,
+or the annotation is unexpectedly missing, WMCO will send a stop signal to the
+kubelet through a new command added to [WMCB](https://github.com/openshift/windows-machine-config-bootstrapper/).
+This new WMCB command, `prepare-upgrade` will send a stop signal to the kubelet
+Windows service, gracefully stopping the kubelet.
+
+Once the Kubelet is stopped, the health check for the Machine associated
```

Comment:
Catching up on MachineHealthChecks it looks like the machine controller attempts to drain the node, but if we stop the kubelet that's not going to happen. Is there another manner in which we could trigger the health check failure?

Reply:
Yes, I believe that's correct let's confirm.

Reply:
We always drain on machine delete regardless of node status/health.  We do this to ensure that PDBs are not violated.  If the kubelet is unreachable, we still use drain to evict pods to ensure PDBs are violated.  For normal healthy kubelets, graceperiods of pods are respected, and we wait for those pods to go away.  For kubelets that are unreachable, we ignore pods with a deletion timestamp greater than 5 minutes.  Unfortunately, the node lifecycle manager will taint the nodes and most pods will be marked deleted due to taints by the scheduler before we even have a chance to drain.  This will invalidate PDBs by itself, so there's not much we can do about it since we ignore pods marked as deleted on unreachable nodes.

Reply:
Okay, so I've now reviewed what's actually being proposed here rather than just responding with how drain works.

This proposal is unsuitable.  Stopping the kubelet is definitely not what we want to do.  That won't allow pods to stop cleanly.

MHC is also an opt-in component.  I'm hesitant to start building in these types of use cases because it's quickly becoming non-optional, and the configuration is user-driven as to what types of things it responds to.

We're already intending to build something into MHC for this kind of thing for spot instances, so the scope creep might already cover what you're trying to do as far as responding to something other than an unhealthy condition.  If you intend to stop the kubelet, you need to ensure the node is cordoned and drained first.

Frankly, none of this is ideal.  Is it possible we can just schedule some MCD to run on these hosts, and just shutdown the host instead of reboot?  This would allow the MCO to proceed through upgrades almost normally, and the MCO already drains things.

I would like to investigate what's possible via the MCO flow first rather than coming to rely on MHC for this.

Reply:
I have changed this as result of the Platform & Lifecycle Architecture call. MHC will no longer be used, and instead the Machine object will be deleted